### ISSUE TITLE
This is not a BUG but I really need some help.

### ISSUE THREAD

**Issue by HuXinjing** [huggingface/accelerate#3336]
### System Info

```Shell
I didn't set accelerate env, just uesed the default setting.
```

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I tried to write a script to run two generate function of LLM with accelerate, but when I load data I found something strange, the length of dataloader before sent to accelerator.prepare(model, dataloader) was 6, but after the function performed, I got len(dataloader)==3, what happened there. It has take my whole day to try to figure it out, but I failed. T_T

Here is part of my log:
Process 0: Dataset length: 5
Process 0: DataLoader length: 5
Process 1: Dataset length: 6
Process 1: DataLoader length: 6
Process 0: DataLoader length after prepare: 3
Process 0: Expected total assistant messages: 1606
# There should Prrocess 1：DataLoader length after prepare: xxx but it was missing..

mycode is like:
    start_idx = (len(eval_data) // accelerator.num_processes) * accelerator.process_index
    end_idx = start_idx + (len(eval_data) // accelerator.num_processes)
    if accelerator.process_index == accelerator.num_processes - 1:
        end_idx = len(eval_data)
    process_eval_data = eval_data[start_idx:end_idx]
    print(f"Process {accelerator.process_index}: Total eval_data samples: {len(process_eval_data)}")
    print(f"Process {accelerator.process_index}: Processing data from index {start_idx} to {end_idx}")
    if merge_weights and base_model_path:
        model, tokenizer = apply_lora(base_model_path, lora_path, accelerator)
    
    model.eval()
    dataset = EvalDataset(process_eval_data)
    print(f"Process {accelerator.process_index}: Dataset length: {len(dataset)}")
    dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=False, collate_fn=collate_fn)
    print(f"Process {accelerator.process_index}: DataLoader length: {len(dataloader)}")
    model, dataloader = accelerator.prepare(model, dataloader)       
    print(f"Process {accelerator.process_index}: DataLoader length after prepare: {len(dataloader)}")

my dataset is a jsonl which have 11 json.  Just lik [{"source":source, "messages":[{message0},{message1}...]},...], each json has different lenth.

I uesed default setting of accelerate and such bash commend to start my script:
    accelerate launch --multi_gpu\
        --num_processes 2 \
        --num_machines 1 \
        --mixed_precision no \
        --dynamo_backend no \
        evaluator_ddp.py \
        --data_path "$DATA_PATH" \
        --fold_idx "$fold" \
        --max_length 23500 \
        --merge_weights True \
        --base_model_path "/root/autodl-tmp/Qwen" \
        --batch_size 1  

### Expected behavior

I think   len(process0.dataloader) + len(process1.dataloader) should be 11.

**Comment by muellerzr** [comment]
Yes, that's normal. We split the dataset by GPU numbers for efficiency.

**Comment by HuXinjing** [comment]
> Yes, that's normal. We split the dataset by GPU numbers for efficiency.

Thanks for your reply. So could I name the specific subset of my dataloader for specific GPU?   For example, for process0, load [item0,2,4,..n], for process1, load the other. Would accelerate split the items with their lenth automatically?

**Comment by muellerzr** [comment]
No. We just evenly distribute them as best we can under the hood

**Comment by HuXinjing** [comment]
> No. We just evenly distribute them as best we can under the hood

Got it. Thanks a lot.
